### ISSUE TITLE
Datastore for etsi-qkd-sdn-node

### ISSUE THREAD

**Issue by alessiocnit** [sysrepo/sysrepo#2957]
I'm trying to use the models:
- etsi-qkd-node-types
- etsi-qkd-sdn-node

I managed to install the modules using the command:
> sysrepoctl --install --yang=etsi-qkd-node-types --owner=root:root --permissions=666

Now I would push datastore into sysrepo. I'm trying to use the following command:
> sysrepocfg --datastore=startup --import=init-qkd-node.xml --format=xml etsi-qkd-sdn-node

But I'm obtaining many errors. I'm actually lost beacuse I don't know if the problem is in the model, in the xml file or in the sysrepo environment. 

I used the --datastore=startup option because otherwise I got the error
- "Cannot operate on the running datastore for ... as there are no active subscriptions for it

I have generated the init-qkd-node.xml using pyang, but the import message generate the following errors
- [ERR] Unable to parse the input data: Failed to resolve identityref "TRANSDUCER". (/etsi....)

**Comment by alessiocnit** [comment]
This is the xml generated with 

>  pyang -f sample-xml-skeleton --sample-xml-skeleton-defaults -o init-qkd-node.xml ./etsi-qkd-sdn-node.yang

`  <sdqkd_node xmlns="urn:etsi:qkd:yang:etsi-qkd-node">
    <node_id>10.100.101.31</node_id>
    <location_id></location_id>
    <qkd_capabilities>
      <link_stats_support>true</link_stats_support>
      <application_stats_support>true</application_stats_support>
      <key_relay_mode_enable>true</key_relay_mode_enable>
    </qkd_capabilities>
    <qkd_interfaces>
      <qkd_interface>
        <qkdi_id>1</qkdi_id>
        <capabilities>
          <role_support>TRANSDUCER</role_support>
          
          <wavelength_range>1-40</wavelength_range>
          <max_absorption>40</max_absorption>
        </capabilities>
        <model>quancom/quantum-node-version-1</model>
        <type>DV-QKD</type>
        
        <att_point>
          <device/>
          <port/>
        </att_point>
      </qkd_interface>
      <qkd_interface>
        <qkdi_id>2</qkdi_id>
        <capabilities>
          <role_support>TRANSDUCER</role_support>
          
          <wavelength_range>1-40</wavelength_range>
          <max_absorption>40</max_absorption>
        </capabilities>
`

**Comment by michalvasko** [comment]
```
"Cannot operate on the running datastore for ... as there are no active subscriptions for it
```
This tells me you are using some ancient version of *sysrepo* so my advice is for you to use a current version, the latest release, for example. I cannot help you with problems in the version you are using.

**Comment by alessiocnit** [comment]
Dear michalvasko, 

I did some progess, now I'm running last version of netopeer2, sysrepo, ...

I managed to properly import:
- etsi-qkd-node-types

using the command:

> sysrepoctl --install ./etsi-... --owner root --group root --permissions 666

For the istallation of:
- etsi-qkd-sdn-node

I'm trying this command:

> sysrepoctl --install ./etsi-... --init-data ./init-etsi.xml 

It seems nice, I used this command and the error messages generated to gradually fill-up the datastore. Finally I got this last error message:

**Comment by alessiocnit** [comment]
This is the datastore I'm trying to use:

  <qkd_node xmlns="urn:etsi:qkd:yang:etsi-qkd-node" xmlns:types="urn:etsi:qkd:yang:etsi-qkd-node-types">
    <qkdn_id>f81d4fae-7dec-11d0-a765-aaaaaaaaaaaa</qkdn_id>
    <qkdn_status>types:NEW</qkdn_status>
    
    <qkdn_version>2.1.1</qkdn_version>
    <qkdn_location_id></qkdn_location_id>
    <qkdn_capabilities>
      <link_stats_support>true</link_stats_support>
      <application_stats_support>true</application_stats_support>
      <key_relay_mode_enable>true</key_relay_mode_enable>
    </qkdn_capabilities>
    <qkd_interfaces>
      <qkd_interface>
        
        
        <qkdi_id>1</qkdi_id>
        <qkdi_model>
          
        </qkdi_model>
        <qkdi_type>types:CV-QKD</qkdi_type>
        
        <qkdi_att_point>
          <device>local-device</device>
          <port>1</port>
        </qkdi_att_point>
        <qkdi_capabilities>
          <role_support>types:TRANSCEIVER</role_support>
            
          <wavelength_range>1,12-20,40,50-80</wavelength_range>
            
          <max_absorption>0.0</max_absorption>
        </qkdi_capabilities>
        <qkdi_status>types:ENABLED</qkdi_status>
          
      </qkd_interface>
      <qkd_interface>
        
        
        <qkdi_id>2</qkdi_id>
        <qkdi_model>
          
        </qkdi_model>
        <qkdi_type>types:CV-QKD</qkdi_type>
        
        <qkdi_att_point>
          <device>local-device</device>
          <port>2</port>
        </qkdi_att_point>
        <qkdi_capabilities>
          <role_support>types:TRANSCEIVER</role_support>
            
          <wavelength_range>1,12-20,40,50-80</wavelength_range>
            
          <max_absorption>0.0</max_absorption>
        </qkdi_capabilities>
        <qkdi_status>types:ENABLED</qkdi_status>
          
      </qkd_interface>
    </qkd_interfaces>
    <qkd_links>
      <qkd_link>
        
        
        <qkdl_id>f81d4fae-aaaa-aaaa-aaaa-00a0c91e6bf6</qkdl_id>
        <qkdl_status>types:OFF</qkdl_status>
          
        <qkdl_enable>true</qkdl_enable>
        <qkdl_local>
          <qkdn_id>f81d4fae-7dec-11d0-a765-aaaaaaaaaaaa</qkdn_id>
          <qkdi_id>1</qkdi_id>
        </qkdl_local>
        <qkdl_remote>
          <qkdn_id>f81d4fae-7dec-11d0-a765-bbbbbbbbbbbb</qkdn_id>
          <qkdi_id>1</qkdi_id>
        </qkdl_remote>
        <qkdl_type>types:PHYS</qkdl_type>
          
          
          
        <phys_channel_att>3.0</phys_channel_att>
        <phys_wavelength>57</phys_wavelength>
          
        <phys_qkd_role>types:TRANSMITTER</phys_qkd_role>
          
        <qkdl_performance>
          <expected_consumption>18</expected_consumption>
          <skr>21</skr>
          <eskr>32</eskr>
          <phys_perf>
            
            
            <perf_type>types:SNR</perf_type>
              
            <value>0.1</value>
          </phys_perf>
        </qkdl_performance>
      </qkd_link>
    </qkd_links>
  </qkd_node>

**Comment by alessiocnit** [comment]
`  <qkd_node xmlns="urn:etsi:qkd:yang:etsi-qkd-node" xmlns:types="urn:etsi:qkd:yang:etsi-qkd-node-types">
    <qkdn_id>f81d4fae-7dec-11d0-a765-aaaaaaaaaaaa</qkdn_id>
    <qkdn_status>types:NEW</qkdn_status>
    
    <qkdn_version>2.1.1</qkdn_version>
    <qkdn_location_id></qkdn_location_id>
    <qkdn_capabilities>
      <link_stats_support>true</link_stats_support>
      <application_stats_support>true</application_stats_support>
      <key_relay_mode_enable>true</key_relay_mode_enable>
    </qkdn_capabilities>
    <qkd_interfaces>
      <qkd_interface>
        
        
        <qkdi_id>1</qkdi_id>
        <qkdi_model>
          
        </qkdi_model>
        <qkdi_type>types:CV-QKD</qkdi_type>
        
        <qkdi_att_point>
          <device>local-device</device>
          <port>1</port>
        </qkdi_att_point>
        <qkdi_capabilities>
          <role_support>types:TRANSCEIVER</role_support>
            
          <wavelength_range>1,12-20,40,50-80</wavelength_range>
            
          <max_absorption>0.0</max_absorption>
        </qkdi_capabilities>
        <qkdi_status>types:ENABLED</qkdi_status>
          
      </qkd_interface>
      <qkd_interface>
        
        
        <qkdi_id>2</qkdi_id>
        <qkdi_model>
          
        </qkdi_model>
        <qkdi_type>types:CV-QKD</qkdi_type>
        
        <qkdi_att_point>
          <device>local-device</device>
          <port>2</port>
        </qkdi_att_point>
        <qkdi_capabilities>
          <role_support>types:TRANSCEIVER</role_support>
            
          <wavelength_range>1,12-20,40,50-80</wavelength_range>
            
          <max_absorption>0.0</max_absorption>
        </qkdi_capabilities>
        <qkdi_status>types:ENABLED</qkdi_status>
          
      </qkd_interface>
    </qkd_interfaces>
    <qkd_links>
      <qkd_link>
        
        
        <qkdl_id>f81d4fae-aaaa-aaaa-aaaa-00a0c91e6bf6</qkdl_id>
        <qkdl_status>types:OFF</qkdl_status>
          
        <qkdl_enable>true</qkdl_enable>
        <qkdl_local>
          <qkdn_id>f81d4fae-7dec-11d0-a765-aaaaaaaaaaaa</qkdn_id>
          <qkdi_id>1</qkdi_id>
        </qkdl_local>
        <qkdl_remote>
          <qkdn_id>f81d4fae-7dec-11d0-a765-bbbbbbbbbbbb</qkdn_id>
          <qkdi_id>1</qkdi_id>
        </qkdl_remote>
        <qkdl_type>types:PHYS</qkdl_type>
          
          
          
        <phys_channel_att>3.0</phys_channel_att>
        <phys_wavelength>57</phys_wavelength>
          
        <phys_qkd_role>types:TRANSMITTER</phys_qkd_role>
          
        <qkdl_performance>
          <expected_consumption>18</expected_consumption>
          <skr>21</skr>
          <eskr>32</eskr>
          <phys_perf>
            
            
            <perf_type>types:SNR</perf_type>
              
            <value>0.1</value>
          </phys_perf>
        </qkdl_performance>
      </qkd_link>
    </qkd_links>
  </qkd_node>`

**Comment by michalvasko** [comment]
Based on all the information here the YANG module `etsi-qkd-sdn-node` is wrong. Specifically, the when condition `../qkdl_type = 'PHYS'` obviously refers to an instance-identifier since the value you are using is `types:PHYS`. However, the value does not include any prefix in the condition and hence uses the default one, which is `node` (the module `etsi-qkd-sdn-node`). This does not match the actual value, hence the error. *libyang* even warns about this when you load the module.

**Comment by alessiocnit** [comment]
Ok so I solved editing the etsi-qkd-sdn-node, specifically I have included the prefix so that now the reference to PHYS and VIRT variables appear as:

"../qkdl_type = 'etsi-qkdn-types:PHYS'"

It seems to work, the modules are correctly loaded into sysrepo, and I'm able to retrieve the datastore using netconf_console2. 

The only doubt leaving is that when I use the --init-data commenad I got the following sysrepo output:

[INF] Connection 3 created.
[INF] Module "etsi-qkd-sdn-node" was installed.
[INF] File "etsi-qkd-sdn-node@2022-01-30.yang" was installed.
[INF] No datastore changes to apply.

Is this correct? I have the -v3 flag on.
